### PR TITLE
[Geo] Refactor GeoShapeQueryBuilder to derive from AbstractGeometryQueryBuilder

### PR DESCRIPTION
**Pull request by nknize**
Refactors `GeoShapeQueryBuilder` to derive from a new `AbstractShapeQueryBuilder` that provides common parsing and build logic for spatial shapes. This will allow development of custom shape queries by extending `AbstractShapeQueryBuilder`, preventing the duplication of common shape query logic.

_note: #44715 further deprecates ShapeBuilder to make the QueryBuilder more spatial agnostic and will be included upon merge._

Relates to #40908

### CODE REVIEW

**Comment by nknize** [comment]
#44715 further deprecates ShapeBuilder to make the QueryBuilder more spatial agnostic and will be included upon merge.

**Comment by nknize** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/2

**Comment by imotov** [comment]
It looks good in general but let's get #44715 in first and merge it into this one. I think it should dramatically simplify `GeoShapeQueryBuilder` ideally leaving only legacy-specific logic there. I think the goal should be to move as much non-legacy logic from`GeoShapeQueryBuilder` into `GeometryIndexer` as possible, otherwise it will force us to add the corresponding `XYShape`-specific logic to `XYShapeQueryBuilder` which is part of HL Rest client.

**Comment by nknize** [comment]
I agree on getting #44715 in first (I'm already merging it with this dev branch as part of the PR review).

We're also going to need to be careful about naming because the field type that exposes Lucene's `LatLonShape` is called `geo_shape` and the field type that exposes Lucene's `XYShape` is called `geometry` so we risk getting confused by terminology where lat,lon geometries are called geo and x,y geometries are called geometry. `GeometryIndexer` is a good example, I think we'll need to carefully document that class to make it clear that it's intended for lat,lon geometries (or rename it to something like `GeodeticIndexer` differentiating between Geodetic lat, lon and Geocentric XYZ, or tangential plane XY).

**Comment by imotov** [comment]
I merged  #44715 into master. Thanks for the review!

**Comment by nknize** [comment]
np; and I just merged those changes from master w/ this PR. Let's see how CI responds and I think this will be good to go. Per our team discuss this morning I renamed `AbstractShapeQueryBuilder` to `AbstractGeometryQueryBuilder` to try and keep the naming conventions clearer. Now we have `GeoShapeQueryBuilder` (and soon `ShapeQueryBuilder`) that will derive from this new `AbstractGeometryQueryBuilder`.  I also like how `AbstractGeometryQueryBuilder` now has no dependencies on jts or spatial4j, only `GeoShapeQueryBuilder` does for legacy purposes. It will make it easier to remove those dependencies once we get there.

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java [hunk 1]
```diff
@@ -471,13 +260,41 @@ protected Query doToQuery(QueryShardContext context) {
...
+    /**
+     * Builds JTS shape from a geometry
+     *
+     * This method is needed to handle legacy indices and will be removed when we no longer need to build JTS shapes
+     */
+    private static Shape buildS4J(Geometry geometry) {
+        return geometryToShapeBuilder(geometry).buildS4J();
+    }
+
     private Query getVectorQuery(QueryShardContext context, Geometry queryShape) {
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: I think in the next iteration I am going to move this into `AbstractGeometryIndexer` and make it obtainable from `QueryShardContext`.

**Review by imotov — APPROVED** [review]
LGTM. Thanks!

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java (hunk 1 above)
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: Do you want to also `@deprecate` it in 8.0 (master) in advance of its removal in 9.0?

Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: Oh... oops.. Sorry, you were referring to `getVectorQuery`.

**Pull request merged**